src/cutpruning.jl

```diff
@@ -109,5 +149,172 @@ function is_cut_relevant(model::SPModel, k::Int, Vt::PolyhedralFunction, solver;
     @objective(m, Min, alpha - dot(λ_k, x) - Vt.betas[k])
     solve(m)
     sol = getobjectivevalue(m)
-    return sol < epsilon
+    return (sol < epsilon), getvalue(x)
+end
+
+
+########################################
+# Territory algorithm
+########################################
+
+Territories(ndim) = Territories(0, [], 0, Array{Float64}(0, ndim))
+
+
+""" Update territories with cuts previously computed during backward pass.  """
+function find_territory!(territory, V, states)
+    nc = V.numCuts
+    # get number of new positions to analyse:
+    nx = size(states, 1)
+    nt = nc - territory.ncuts
+
+    for i in 1:nt
+        add_cut!(territory)
+        update_territory!(territory, V, nc - nt + i)
+    end
+
+    # ensure that territory has the same number of cuts as V!
+    assert(territory.ncuts == V.numCuts)
+
+    for i in 1:nx
+        x = collect(states[i, :])
+        add_state!(territory, V, x)
+    end
+
 end
+
+
+"""Update territories considering new cut with index `indcut`."""
+function update_territory!(territory, V, indcut)
+    for k in 1:territory.ncuts
+        if k == indcut
+            continue
+        end
+        todelete = []
+        for (num, (ix, cost)) in enumerate(territory.territories[k])
+            x = collect(territory.states[ix, :])
+
+            costnewcut = cutvalue(V, indcut, x)
+
+            if costnewcut > cost
+                push!(todelete, num)
+                push!(territory.territories[indcut], (ix, costnewcut))
+            end
+        end
+        deleteat!(territory.territories[k], todelete)
+    end
+end
+
+
+"""Add cut to `territory`."""
+function add_cut!(territory)
+    push!(territory.territories, [])
+    territory.ncuts += 1
+end
+
+
+"""Add a new state and update territories."""
+function add_state!(territory::Territories, V::PolyhedralFunction, x::Array{Float64})
+    # Get cut which is the supremum at point `x`:
+    bcost, bcuts = optimalcut(x, V)
+
+    # Add `x` to the territory of cut `bcuts`:
+    territory.nstates += 1
+    push!(territory.territories[bcuts], (territory.nstates, bcost))
+
+    # Add `x` to the list of visited state:
+    territory.states = vcat(territory.states, x')
+end
+
+
+"""Remove empty cuts with heuristic in PolyhedralFunction."""
+function remove_empty_cuts!(territory::Territories, V::PolyhedralFunction)
+    assert(territory.ncuts == V.numCuts)
+
+    nstates = [length(terr) for terr in territory.territories]
+    active_cuts = nstates .> 0
+
+    territory.territories = territory.territories[active_cuts]
+    territory.ncuts = sum(active_cuts)
+    return PolyhedralFunction(V.betas[active_cuts],
+                              V.lambdas[active_cuts, :],
+                              sum(active_cuts))
+end
+
+
+"""Remove empty cuts in PolyhedralFunction with usefulness test."""
```

Comment:
It seems that this is also an exact pruning, but done by solving less LP, is it right ?